src/Features/Core/Portable/Common/AbstractProjectExtensionProvider.cs

```diff
@@ -29,27 +28,44 @@ internal abstract class AbstractProjectExtensionProvider<TProvider, TExtension,
         protected abstract ImmutableArray<string> GetLanguages(TExportAttribute exportAttribute);
         protected abstract bool TryGetExtensionsFromReference(AnalyzerReference reference, out ImmutableArray<TExtension> extensions);
 
+        public static bool TryGetCachedExtensions(IReadOnlyList<AnalyzerReference> analyzerReferences, out ImmutableArray<TExtension> extensions)
+        {
+            if (s_referencesToExtensionsMap.TryGetValue(analyzerReferences, out var providers))
+            {
+                extensions = providers.Value;
+                return true;
+            }
+
+            extensions = ImmutableArray<TExtension>.Empty;
+            return false;
+        }
+
         public static ImmutableArray<TExtension> GetExtensions(Project? project)
         {
             if (project is null)
                 return ImmutableArray<TExtension>.Empty;
 
-            if (s_referencesToExtensionsMap.TryGetValue(project.AnalyzerReferences, out var providers))
-                return providers.Value;
+            return GetExtensions(project.Language, project.AnalyzerReferences);
+        }
+
+        public static ImmutableArray<TExtension> GetExtensions(string language, IReadOnlyList<AnalyzerReference> analyzerReferences)
+        {
+            if (TryGetCachedExtensions(analyzerReferences, out var providers))
+                return providers;
 
-            return GetExtensionsSlow(project);
+            return GetExtensionsSlow(language, analyzerReferences);
 
-            ImmutableArray<TExtension> GetExtensionsSlow(Project project)
-                => s_referencesToExtensionsMap.GetValue(project.AnalyzerReferences, _ => new(ComputeExtensions(project))).Value;
+            ImmutableArray<TExtension> GetExtensionsSlow(string language, IReadOnlyList<AnalyzerReference> analyzerReferences)
+                => s_referencesToExtensionsMap.GetValue(analyzerReferences, _ => new(ComputeExtensions(language, analyzerReferences))).Value;
```

Comment:
So the computation requires the language name but we aren't including that in the original key? This isn't new but is this broken? Or we're just assuming that each project gets its own list of analyzer references which _might_ be the same content wise but just happen to be different objects?

Reply:
Is it not guaranteed that the list of `AnalyzerReference` for each Project is unique object-wise?

Reply:
Oh, probably. That's....interesting. Yikes that we might depend on that. (not new here, but yikes still)